File: leetcode/algorithms/35_search_insert_position/main.cpp
#include <algorithm>
#include <vector>
using namespace std;

class SearchInsertPosition {
public:
    /**
     * Binary Search
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(1)
     */
    int searchInsert(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            if (target > nums[mid]) {
                left = mid + 1;
            } else if (target < nums[mid]) {
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return left;
    }


    // Solution
    /**
     * Solution 1
     * 
     * Binary Search
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(1)
     */
    int solution1(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return left;
    }

    /**
     * Solution 2
     * 
     * algorithm::lower_bound
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(1)
     */
    int solution2(vector<int>& nums, int target) {
        return lower_bound(nums.begin(), nums.end(), target) - nums.begin();
    }
};

File: leetcode/algorithms/35_search_insert_position/main.dart
class SearchInsertPosition {
  /**
   * Binary Search
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(logᴺ)
   *   - Space Complexity: O(1)
   */
  int searchInsert(List<int> nums, int target) {
    for (var i = 0; i < nums.length; i++) {
      if (nums[i] >= target) {
        return i;
      }
    }

    return nums.length;
  }


  // Solution
  /**
   * Binary Search
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(logᴺ)
   *   - Space Complexity: O(1)
   */
  int solution(List<int> nums, int target) {
    int low = 0;
    int high = nums.length - 1;

    while (low <= high) {
      int mid = low + (high - low) ~/ 2;

      if (nums[mid] == target) {
        return mid;
      } else if (nums[mid] < target) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }

    return low;
  }
}
